pkg/providers/vsphere/virtualmachine/snapshot.go

```diff
@@ -0,0 +1,89 @@
+// © Broadcom. All Rights Reserved.
+// The term “Broadcom” refers to Broadcom Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: Apache-2.0
+
+package virtualmachine
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha4"
+	vmopv1common "github.com/vmware-tanzu/vm-operator/api/v1alpha4/common"
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+)
+
+// SnapshotArgs contains the options for createSnapshot.
+type SnapshotArgs struct {
+	VMCtx      pkgctx.VirtualMachineContext
+	VcVM       *object.VirtualMachine
+	VMSnapshot vmopv1.VirtualMachineSnapshot
+}
+
+func SnapshotVirtualMachine(args SnapshotArgs) (*types.ManagedObjectReference, error) {
+	obj := args.VMSnapshot
+	vm := args.VcVM
+	// find snapshot by name
+	snapMoRef, _ := vm.FindSnapshot(args.VMCtx, obj.Name)
+	if snapMoRef != nil {
+		// TODO: Handle revert to snapshot
```

Comment:
We also need to handle the scenario where the spec.currentSnapshot exists on VC VM, but because we only compare it with `vm.Status.CurrentSnapshot` in vmprovider_vm.go L1118, we fell through in this code path.
```
if vmCtx.VM.Spec.CurrentSnapshot == nil || reflect.DeepEqual(vmCtx.VM.Spec.CurrentSnapshot, vmCtx.VM.Status.CurrentSnapshot) {
	return nil
}
```
So, basically compare VM's [currentSnapshot](https://developer.broadcom.com/xapis/vsphere-web-services-api/8.0u3/vim.vm.SnapshotInfo.html?scrollString=currentSnapshot) with the spec.currentSnapshot.

Reply:
Hmm. The VM.SnapshotInfo.CurrentSnapshot MoRef can give us the MOID of the snapshot but our spec.currentSnapshot would be still the name (not the MOID). So, We don't have a way to compare the unique vcenter MOIDs. 

Currently the way this is handled since revert is a TODO. When the snapshot is already present, it just patches the status of the VM. I agree though, The problem you mentioned would arise, when someone deletes the exists snapshot CR and creates a new one with the same name. This might have to be punted for when we handle RevertToSnapshot() here. Let me add a TODO if that sounds alright.

Reply:
Yeah, TODO is fine for now.  But it would be simple.  You can get the name of the snapshot from the Moref of the snapshot and then compare that with the name of the snapshot object's name.

Reply:
> You can get the name of the snapshot from the Moref of the snapshot and then compare that with the name of the snapshot object's name

ah. I thought I could get the name too. But from what I found, the MoRef gives access to a ref.Value() (the MOID "snap-01" or similar) and the ref.Type() ("VirtualMachineSnapshot") and the ServerGUID() . I wasn't sure we can fetch the name from here?. I've added TODO to handle this as a follow-up.